hazelcast/src/main/java/com/hazelcast/config/Config.java

```diff
@@ -237,9 +237,24 @@ private static Config applyEnvAndSystemVariableOverrides(Config cfg) {
         cfg = new ExternalConfigurationOverride().overwriteMemberConfig(cfg);
         PersistenceAndHotRestartPersistenceMerger
                 .merge(cfg.getHotRestartPersistenceConfig(), cfg.getPersistenceConfig());
+        setConfigurationFileFromUrl(cfg);
         return cfg;
     }
 
+    // configurationFile must be set correctly because dynamic
+    // configuration persistence depends on this field. If this is
+    // absent, hazelcast instance may fail to find a file to persist.
+    private static void setConfigurationFileFromUrl(Config cfg) {
+        if (cfg.getConfigurationFile() == null && cfg.getConfigurationUrl() != null) {
+            File configFile = new File(cfg.getConfigurationUrl().getPath());
+
+            // Only set configurationFile if the config actually exist on the filesystem.
+            if (configFile.exists()) {
+                cfg.setConfigurationFile(configFile);
+            }
```

Comment:
What do you think about adding a warning when the `configurationFile` is set to a nonexisting file?

Reply:
It may be intentional. For example people may want to embed files in JARs. In that case would it be preferable to still warn?

Reply:
I see, fair point.